File: 98.ValidateBinarySearchTree/98.ValidateBinarySearchTree.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct{
	Val int
	Left *TreeNode
	Right * TreeNode
}

func main(){
	root := &TreeNode{5, nil, nil}
	root.Left = &TreeNode{3, nil, nil}
	root.Right = &TreeNode{7, nil, nil}
	root.Right.Left = &TreeNode{4, nil, nil}
	root.Right.Right = &TreeNode{8, nil, nil}  

	fmt.Println(isValidBST(root))
}

func isValidBST(root *TreeNode) bool {
	return dps(root, math.MinInt, math.MaxInt)
}

func dps(root *TreeNode, lower int, upper int) bool {
	if root == nil { 
		return true 
	}
	if lower >= root.Val || root.Val >= upper {
		return false
	}
	return dps(root.Left, lower, root.Val ) && dps(root.Right, root.Val, upper)
}

File: 98.ValidateBinarySearchTree/98.ValidateBinarySearchTree.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

let root = new TreeNode(5);
root.left = new TreeNode(3);
root.right = new TreeNode(7);

root.right.left = new TreeNode(4);
root.right.right = new TreeNode(8);

const isValidBST = (root) => {
  const dps = (root, lower, upper) => {
    if (!root) return true;
    if (lower > root.val || root.val > upper) {
      return false;
    }
    return dps(root.left, lower, root.val) && dps(root.right, root.val, upper);
  };

  return dps(root, -Infinity, Infinity);
};

console.log(isValidBST(root));
